mayastor/src/bdev/nexus/nexus_metadata.rs

```diff
@@ -5,821 +5,745 @@
-            entry_size: MetaDataHeader::INDEX_ENTRY_SIZE,
-            index_checksum: 0,
-            data_start: data_start as u64,
-            data_end: partition.ent_end - partition.ent_start - 1,
+            timestamp: 0,
+            table_crc: 0,
+            children: Vec::new(),
         }
     }
 }
 
-#[derive(Debug, Deserialize, PartialEq, Default, Serialize, Clone)]
-pub struct MetaDataIndexEntry {
-    /// Current object revision
-    pub revision: u64,
-    /// When this entry was created (number of microseconds since UNIX_EPOCH)
-    pub timestamp: u128,
-    /// CRC-32 checksum of all the blocks comprising the stored object
-    pub data_checksum: u32,
-    /// Location of first data block (relative to self_lba in the
-    /// MetaDataHeader)
-    pub data_start: u64,
-    /// Location of last data block (relative to self_lba in the
-    /// MetaDataHeader)
-    pub data_end: u64,
-}
-
-impl MetaDataIndexEntry {
-    /// Convert a slice into an index array
-    pub fn from_slice(
-        slice: &[u8],
-        used_entries: u32,
-    ) -> Result<Vec<MetaDataIndexEntry>, MetaDataError> {
-        let mut index: Vec<MetaDataIndexEntry> = Vec::new();
-        let mut reader = Cursor::new(slice);
-        for _ in 0 .. used_entries {
-            index.push(
-                deserialize_from(&mut reader).context(DeserializeError {})?,
-            );
-        }
-        Ok(index)
+impl Default for MetaDataObject {
+    fn default() -> Self {
+        Self::new()
     }
+}
 
-    /// Calculate checksum of an index array
-    pub fn checksum(index: &[MetaDataIndexEntry]) -> u32 {
-        let mut digest = crc32::Digest::new(crc32::IEEE);
-        for entry in index {
-            digest.write(&serialize(entry).unwrap());
+impl From<&MetaDataObject> for MetaDataHeader {
+    fn from(object: &MetaDataObject) -> MetaDataHeader {
+        MetaDataHeader {
+            signature: object.signature,
+            self_checksum: object.self_checksum,
+            generation: object.generation,
+            timestamp: object.timestamp,
+            table_crc: object.table_crc,
+            count: object.children.len() as u32,
         }
-        digest.sum32()
     }
 }
 
-#[derive(Debug, PartialEq, Default, Clone)]
-pub struct NexusMetaData {
-    pub header: MetaDataHeader,
-    pub index: Vec<MetaDataIndexEntry>,
+#[derive(Debug, Deserialize, Serialize)]
+pub struct MetaDataIndex {
+    pub signature: [u8; 8],
+    pub index_size: u32,
+    pub self_checksum: u32,
+    pub generation: u64,
+    pub timestamp: u128,
+    pub self_lba: u64,
+    pub start_lba: u64,
+    pub current_entry: u64,
+    pub used_entries: u64,
+    pub total_entries: u64,
 }
 
-impl NexusMetaData {
-    /// Construct a MetaDataHeader from raw data
-    fn read_header(buf: &DmaBuf) -> Result<MetaDataHeader, MetaDataError> {
-        MetaDataHeader::from_slice(buf.as_slice())
-    }
+impl MetaDataIndex {
+    pub const METADATA_INDEX_SIZE: u32 = 80;
```

Comment:
Would't it be better to express this constant via the size of the structure ?

Reply:
Yes. Done.